### PR TITLE
Add UCXX support

### PR DESCRIPTION
**Pull request by pentschev**
Add support for [UCXX](https://github.com/rapidsai/ucxx). It is our intention to soon switch from UCX-Py to UCXX and archive the former.

This PR adds support for UCXX on the C++ backend but retains the original UCX implementation for now (based on the UCP layer), moving to UCXX will simplify RAFT code a bit given the UCXX implementation requires fewer lines of boilerplate code.

On the Python front raft-dask tests are added for both UCX-Py (which there weren't any) and UCXX. The UCX-Py tests continue to use the UCX (UCP layer) implementation, whereas the UCXX tests use the UCXX C++ implementation.

When the switch is complete we can remove all previous UCX/UCX-Py code from the RAFT codebase. If for some reason using the UCX (UCP layer) is preferred on the C++ backend instead of the UCXX C++ implementation this is possible, but UCX-Py code will be archived and dropped in favor of the UCXX Python backend.

### CODE REVIEW

**Comment by cjnolet** [comment]
@pentschev are you ready for this to be merged by 24.02?

**Comment by pentschev** [comment]
> @pentschev are you ready for this to be merged by 24.02?

I'm hoping for that but we still need UCXX wheels which is still ongoing work. I can't say whether they'll be ready yet. On the UCXX side this PR is working as expected locally so if we're able to get wheels built I think we won't have much trouble getting this in.

**Comment by vyasr** [comment]
It looks like ucx's config is unfortunately not idempotent. We'll need to implement a workaround for that. Otherwise everything is looking solid though.

**Comment by pentschev** [comment]
> It looks like ucx's config is unfortunately not idempotent. We'll need to implement a workaround for that. Otherwise everything is looking solid though, pending tests of raft-dask wheels to see how well ucxx works at runtime.

@vyasr Can you elaborate in layman's terms what that means in this context? Is this something we can fix in UCXX or not possible given the need/use of specific properties in UCXX?

Thanks so much for the help in pushing this forward!

**Comment by vyasr** [comment]
I'm not yet sure how to fix this in UCXX, will think on it this morning. There's probably some way it can be done. The longer-term fix really ought to be in UCX itself.

The failure in conda Python builds is fixable in UCXX's CMake.

**Comment by pentschev** [comment]
Since we got blocked on releasing UCXX wheel for 24.02 this cannot make it either, retargeting to 24.04.

**Comment by vyasr** [comment]
OK, I was able to fix the issues by modifying UCXX. We'll eventually need to undo those fixes once UCX is patched (see https://github.com/rapidsai/ucxx/issues/173), but until then what we have here should work.

**Comment by pentschev** [comment]
> OK, I was able to fix the issues by modifying UCXX. We'll eventually need to undo those fixes once UCX is patched (see [rapidsai/ucxx#173](https://github.com/rapidsai/ucxx/issues/173)), but until then what we have here should work.

Thanks a lot Vyas! Those changes probably won't make their way in until UCX 1.17.x, maybe 1.16.x but I'll have to check. So I'll keep an eye on that and revert when we move to newer UCX.

**Review by AyodeAwe — APPROVED** [review]

**Comment by quasiben** [comment]
@cjnolet is it too late to review this for 24.04 ?  We'll retarget for 24.06 but we'd still like your review before merging in

**Review by bdice — COMMENTED** [review]
I only reviewed the CMake and versioning/dependency updates. They seem fine to me, but there are pending conflicts to resolve in `ci/release/update-version.sh` and `python/raft-dask/CMakeLists.txt`. This should probably be retargeted to 24.06, since the 24.04 release is happening this week, so I'll hold off on approving until the versions are changed.

**Comment by pentschev** [comment]
> I only reviewed the CMake and versioning/dependency updates. They seem fine to me, but there are pending conflicts to resolve in `ci/release/update-version.sh` and `python/raft-dask/CMakeLists.txt`. This should probably be retargeted to 24.06, since the 24.04 release is happening this week, so I'll hold off on approving until the versions are changed.

Thanks @bdice , branch was retargeted, conflicts are resolved and versions updated. 🙂

**Comment by pentschev** [comment]
The CI failures are not problems with the PR but rather recent changes in `rapids-dask-dependency`/`dask-cuda`. We're discussing how to better address it (short-term probably means adding `dask[dataframe]` as a `dask-cuda` dependency) and we should have a solution to make tests pass by tomorrow.

**Comment by cjnolet** [comment]
I think these changes look fine but the real telling piece will be if cuML and cugraph both pass CI with the change.

**Comment by pentschev** [comment]
> I think these changes look fine but the real telling piece will be if cuML and cugraph both pass CI with the change.

Agreed, once this PR is merged then https://github.com/rapidsai/cuml/pull/5697 can also be merged so we can have a better overview of how things are looking.

**Comment by pentschev** [comment]
The error is now gone and all tests are passing again, so should be good for a review. @divyegala mentioned he would do that, no absolutely immediate pressure but would be good if we could have this merged some time this week or next the latest.

**Review by divyegala — APPROVED** [review]

File: ci/test_wheel_raft_dask.sh
Comment (vyasr) [inline ci/test_wheel_raft_dask.sh:9]:
```suggestion
ucxx_wheelhouse=$(RAPIDS_PY_WHEEL_NAME="ucxx_${RAPIDS_PY_CUDA_SUFFIX}" rapids-get-pr-wheel-artifact ucxx 226 python)
```

File: ci/test_wheel_raft_dask.sh
Comment (vyasr) [inline ci/test_wheel_raft_dask.sh:9]:
```suggestion
ucxx_wheelhouse=$(RAPIDS_PY_WHEEL_NAME="ucxx_${RAPIDS_PY_CUDA_SUFFIX}" rapids-get-pr-wheel-artifact ucxx 226 python)
distributed_ucxx_wheelhouse=$(RAPIDS_PY_WHEEL_NAME="distributed_ucxx_${RAPIDS_PY_CUDA_SUFFIX}" rapids-get-pr-wheel-artifact ucxx 226 python)
```

File: ci/test_wheel_raft_dask.sh
Comment (vyasr) [inline ci/test_wheel_raft_dask.sh:16]:
```suggestion
python -m pip install "raft_dask-${RAPIDS_PY_CUDA_SUFFIX}[test]>=0.0.0a0" --find-links dist/ --find-links "${ucxx_wheelhouse}" --find-links "${distributed_ucxx_wheelhouse}"
```

File: conda/recipes/raft-dask/meta.yaml
Comment (vyasr) [inline conda/recipes/raft-dask/meta.yaml:1]: The `ucx` version specification in here isn't quite right. The `conda_build_config.yaml` file has the looser version pinning, so we'll always end up pulling 1.16 to build against as well as run. I think we want to always build against the oldest right? So we need to specify `==1.15` for the build.

File: ci/test_wheel_raft_dask.sh
Comment (vyasr) [inline ci/test_wheel_raft_dask.sh:23]: These flag names are a bit confusing. Do they really mean only run the ucx-py/ucxx tests? What happens if you don't pass either flag? Do _all_ tests run?

**Review by vyasr — CHANGES_REQUESTED** [review]
One change on the ucx dependency in conda, otherwise I'm generally fine with getting this PR merged.

File: ci/test_wheel_raft_dask.sh
Comment (pentschev) [reply ci/test_wheel_raft_dask.sh:23]: Passing no flags run all Dask tests with `protocol="tcp"`, including those that are not specifically meant to test comms. With `--run_ucx`/`run_ucxx` _only_ tests specifically mean to test comms run with `protocol="ucx"`(UCX-Py)/`protocol="ucxx"`(UCXX)`. The reason for running them separately is that we need to create full Dask clusters with the respective comms configuration, setting clusters up take both time and, more importantly, memory, thus we cannot run all in one pass.

File: conda/recipes/raft-dask/meta.yaml
Comment (pentschev) [reply conda/recipes/raft-dask/meta.yaml:1]: As discussed offline, I don't think it truly makes sense to explicitly include `ucx` in the conda package as `raft-dask` will not be used without `ucx-py` or `ucxx` which will both bring `ucx` transitively, and thus reduce one more place where we have to explicitly synchronize UCX versions in the RAPIDS stack.

**Review by vyasr — APPROVED** [review]

**Pull request merged**
